Source/EasyNetQ/PullingConsumer.cs

```diff
@@ -0,0 +1,435 @@
+        }
+
+        /// <summary>
+        ///     True is a message is available
+        /// </summary>
+        public bool IsAvailable => isAvailable;
+
+        /// <summary>
+        ///     Returns remained messages count if the message is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public ulong RemainedMessagesCount
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return messagesCount;
+            }
+        }
+
+        /// <summary>
+        ///     Returns received info if the message is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public MessageReceivedInfo ReceivedInfo
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return receivedInfo;
+            }
+        }
+
+        /// <summary>
+        ///     Returns message if it is available
+        /// </summary>
+        /// <exception cref="InvalidOperationException"></exception>
+        public IMessage<T> Message
+        {
+            get
+            {
+                if (!isAvailable)
+                    throw new InvalidOperationException("No message is available");
+
+                return message;
+            }
+        }
+    }
+
+    /// <summary>
+    ///     Allows to receive messages by pulling them one by one
+    /// </summary>
+    public interface IPullingConsumer : IDisposable
+    {
+        /// <summary>
+        ///     Receives a single message
+        /// </summary>
+        /// <param name="cancellationToken">The cancellation token</param>
+        /// <returns></returns>
+        Task<PullResult> PullAsync(CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///     Acknowledges one or more messages
+        /// </summary>
+        /// <param name="deliveryTag"></param>
+        /// <param name="multiple"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        Task AckAsync(ulong deliveryTag, bool multiple, CancellationToken cancellationToken = default);
+
+        /// <summary>
+        ///     Rejects one or more messages
+        /// </summary>
+        /// <param name="deliveryTag"></param>
+        /// <param name="multiple"></param>
+        /// <param name="requeue"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        Task RejectAsync(ulong deliveryTag, bool multiple, bool requeue, CancellationToken cancellationToken = default);
+    }
+
+
+    /// <summary>
+    ///     Allows to receive messages by pulling them one by one
+    /// </summary>
+    public interface IPullingConsumer<T> : IDisposable
```

Comment:
Consider inherit `IPullingConsumer<T>` from `IPullingConsumer`. `AckAsync` and `RejectAsync` have identical signatures. I see that `PullAsync` signatures differ only by return type. Well, it is not a big deal as you can use `new` modifier.